### PR TITLE
Add "offline" data cache generation support

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?

1. Adds "offline" data cache generation support. The training script will be stopped once data cache is generated.
2. Enable mcore's `num_dataset_builder_threads` param usage.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 1]
```diff
@@ -1662,6 +1663,12 @@ def setup(self, stage=None):
             # Override limit_val_batches to be a multiple of num microbatches to prevent val_step from exiting in between a step
             self._reconfigure_limit_batches(self.trainer.limit_val_batches, self._validation_dl, 'val')
 
+        # Data cache generation only
+        # Stops training after creating a data cache
+        if self.cfg.data.get('data_cache_generation_only', False):
+            self.trainer.num_sanity_val_steps = 0
+            self.trainer.should_stop = True
+
```
Comment (athitten) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1671]: @dimapihtar training wouldn't have started at all right to stop it here ? Since we are checking in the setup() method which is run before training begins.

Comment (dimapihtar) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1671]: > @dimapihtar training wouldn't have started at all right to stop it here ? Since we are checking in the setup() method which is run before training begins.

yes, but trainer is already initialized at that moment so it quits the scripts.

Comment (dimapihtar) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1671]: @athitten it quits the script before the first training step.

**Comment by samskillman** [comment]
What is the reasoning behind overloading num_workers for the num_dataset_builder_threads, instead of passing num_dataset_builder_threads in as a separate config parameter? 

I could imagine a case where I set num_dataset_builder_threads to a very large number (64 or 128) but wouldn't want that many workers during the typical data loading. That would allow many workloads to specify a high num_dataset_builder_threads without needing to explicitly use the data_cache_generation_only offline step.

Instead I would suggest num_dataset_builder_threads as an explicit new parameter in megatron_gpt_config.yaml, and default it to 1 to mirror https://github.com/NVIDIA/Megatron-LM/blob/e33c8f78a35765d5aa37475a144da60e8a2349d1/megatron/core/datasets/blended_megatron_dataset_config.py#L48

**Comment by dimapihtar** [comment]
> What is the reasoning behind overloading num_workers for the num_dataset_builder_threads, instead of passing num_dataset_builder_threads in as a separate config parameter?
> 
> I could imagine a case where I set num_dataset_builder_threads to a very large number (64 or 128) but wouldn't want that many workers during the typical data loading. That would allow many workloads to specify a high num_dataset_builder_threads without needing to explicitly use the data_cache_generation_only offline step.
> 
> Instead I would suggest num_dataset_builder_threads as an explicit new parameter in megatron_gpt_config.yaml, and default it to 1 to mirror https://github.com/NVIDIA/Megatron-LM/blob/e33c8f78a35765d5aa37475a144da60e8a2349d1/megatron/core/datasets/blended_megatron_dataset_config.py#L48

thank you for the review, fixed.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 1 above)
Comment (athitten) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1692]: Thanks @dimapihtar, the comment `# Stops training after creating a data cache` made me think training would have started. So with `data_cache_generation_only` arg we just cache data and don't run or start any training at all ?

Comment (dimapihtar) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1692]: @athitten yes, exactly. Probably not the best explanation from my side.

**Review by samskillman — APPROVED** [review]

**Review by athitten — APPROVED** [review]
LGTM, thanks @dimapihtar !
